core/include/traccc/ambiguity_resolution/greedy_ambiguity_resolution_algorithm.hpp

```diff
@@ -0,0 +1,161 @@
+#include "traccc/edm/track_candidate.hpp"
+#include "traccc/edm/track_state.hpp"
+#include "traccc/utils/algorithm.hpp"
+
+// Greedy ambiguity resolution adapted from ACTS code
+
+namespace traccc {
+
+/// Evicts tracks that seem to be duplicates or fakes. This algorithm takes a
+/// greedy approach in the sense that it will remove the track which looks "most
+/// duplicate/fake" first and continues the same process with the rest. That
+/// process continues until the final state conditions are met.
+///
+/// The implementation works as follows:
+///  1) Calculate shared hits per track.
+///  2) If the maximum shared hits criteria is met, we are done.
+///     This is the configurable amount of shared hits we are ok with
+///     in our experiment.
+///  3) Else, remove the track with the highest relative shared hits (i.e.
+///     shared hits / hits).
+///  4) Back to square 1.
+class greedy_ambiguity_resolution_algorithm
+    : public algorithm<track_state_container_types::host(
+          const typename track_state_container_types::host&)> {
+
+    public:
+    struct config_t {
+        /// Maximum amount of shared hits per track. One (1) means "no shared
+        /// hit allowed".
+        std::uint32_t maximum_shared_hits = 1;
+
+        /// Maximum number of iterations.
+        std::uint32_t maximum_iterations = 1000000;
+
+        /// Minimum number of measurement to form a track.
+        std::size_t n_measurements_min = 3;
+
+        // True if obvious errors should be checked after the completion
+        // of the algorithm.
+        bool check_obvious_errs = true;
+
+        bool verbose_info = true;
+        bool verbose_error = true;
+        bool verbose_flood = false;
+    };
+
+    struct state_t {
+        std::size_t number_of_tracks{};
+
+        /// For this whole comment section, track_index refers to the index of a
+        /// track in the initial input container.
+        ///
+        /// There is no (track_id) in this algorithm, only (track_index).
+
+        /// Associates each track_index with the track's chi2 value
+        std::vector<traccc::scalar> track_chi2;
+
+        /// Associates each track_index to the track's (measurement_id)s list
+        std::vector<std::vector<std::size_t>> measurements_per_track;
+
+        /// Associates each measurement_id to a set of (track_index)es sharing
+        /// it
+        std::unordered_map<std::size_t, std::set<std::size_t>>
+            tracks_per_measurement;
+
+        /// Associates each track_index to its number of shared measurements
+        /// (among other tracks)
+        std::vector<std::size_t> shared_measurements_per_track;
+
+        /// Keeps the selected tracks indexes that have not (yet) been removed
+        /// by the algorithm
+        std::set<std::size_t> selected_tracks;
+    };
+
+    void verbose_info(std::string s) const;
+
+    void verbose_error(std::string s) const;
+
+    void verbose_flood(std::string s) const;
```

Comment:
You're absolutely right, I'm sorry for this mediocre piece of code. The output of the ACTS Greedy Solver [uses the macro `ACTS_VERBOSE`](https://github.com/acts-project/acts/blob/d0432987602b759959a203a7c47f5c1b5cc84bd0/Core/src/AmbiguityResolution/GreedyAmbiguityResolution.cpp#L81), which itself calls [the macro `ACTS_LOG`](https://github.com/acts-project/acts/blob/d0432987602b759959a203a7c47f5c1b5cc84bd0/Core/include/Acts/Utilities/Logger.hpp#L66) and outputs to a stream, when the log level matches the requirement of the current logger.

I briefly looked for a proper logger in traccc but haven’t found any. Does traccc have a logger? Otherwise I will use the same *stream* method used by ACTS.

Reply:
We don't have it in traccc.  @niermann999 might have an idea on this.

Of course we can improve these things later